cpp/src/arrow/engine/expression.h

```diff
@@ -0,0 +1,490 @@
+  explicit MultiAryOpMixin(std::vector<std::shared_ptr<Expr>> operands)
+      : operands_(std::move(operands)) {}
+
+  std::vector<std::shared_ptr<Expr>> operands_;
+};
+
+///
+/// Value Expressions
+///
+
+/// An unnamed scalar literal expression.
+class ARROW_EN_EXPORT ScalarExpr : public Expr {
+ public:
+  static Result<std::shared_ptr<ScalarExpr>> Make(std::shared_ptr<Scalar> scalar);
+
+  const std::shared_ptr<Scalar>& scalar() const { return scalar_; }
+
+ private:
+  explicit ScalarExpr(std::shared_ptr<Scalar> scalar);
+
+  std::shared_ptr<Scalar> scalar_;
+};
+
+/// References a column in a table/dataset
+class ARROW_EN_EXPORT FieldRefExpr : public UnaryOpMixin, public Expr {
+ public:
+  static Result<std::shared_ptr<FieldRefExpr>> Make(std::shared_ptr<Expr> input,
+                                                    int index);
+  static Result<std::shared_ptr<FieldRefExpr>> Make(std::shared_ptr<Expr> input,
+                                                    std::string field_name);
+
+  int index() const { return index_; }
+
+ private:
+  FieldRefExpr(std::shared_ptr<Expr> input, int index);
+
+  int index_;
+};
+
+///
+/// Comparison expressions
+///
+
+class ARROW_EN_EXPORT CompareOpExpr : public BinaryOpMixin, public Expr {
+ public:
+  CompareKind compare_kind() const { return compare_kind_; }
+
+  /// This inner-class is required because `using` statements can't use derived
+  /// methods.
+  template <typename Derived>
+  struct MakeMixin {
+    static Result<std::shared_ptr<Derived>> Make(std::shared_ptr<Expr> left,
+                                                 std::shared_ptr<Expr> right) {
+      if (left == NULLPTR || right == NULLPTR) {
+        return Status::Invalid("Compare operands must be non-nulls");
+      }
+
+      // Broadcast the comparison to the biggest shape.
+      ARROW_ASSIGN_OR_RAISE(auto broadcast,
+                            ExprType::Broadcast(left->type(), right->type()));
+      // And change this shape's type to boolean.
+      ARROW_ASSIGN_OR_RAISE(auto type, broadcast.WithType(boolean()));
+
+      return std::shared_ptr<Derived>(new Derived(std::move(type),
+                                                  expr_traits<Derived>::compare_kind_id,
+                                                  std::move(left), std::move(right)));
+    }
+  };
+
+ protected:
+  CompareOpExpr(ExprType type, CompareKind op, std::shared_ptr<Expr> left,
+                std::shared_ptr<Expr> right)
+      : BinaryOpMixin(std::move(left), std::move(right)),
+        Expr(COMPARE_OP, std::move(type)),
+        compare_kind_(op) {}
+
+  CompareKind compare_kind_;
+};
+
+template <typename Derived>
+class BaseCompareExpr : public CompareOpExpr,
+                        protected CompareOpExpr::MakeMixin<Derived> {
+ public:
+  using CompareOpExpr::MakeMixin<Derived>::Make;
+
+ protected:
+  using CompareOpExpr::CompareOpExpr;
+};
+
+class ARROW_EN_EXPORT EqualExpr : public BaseCompareExpr<EqualExpr> {
```

Comment:
Given an `Expr` (e.g. `std::shared_ptr<Expr>`) that is actually an `EqualExpr`, how does one determine that it is `EqualExpr`?

Reply:
I'll make some utility in type_traits for this. LLVM has a suite of [utilities](http://llvm.org/docs/ProgrammersManual.html#the-isa-cast-and-dyn-cast-templates) that I find handy for such things